New User Experience and Enhancement posts.

dero opened the pull request:
Supersedes #170 

## Summary

- Adds two new categories: **User Experience** and **Enhancements**.
- Each category contains 3 new posts explaining the technical additions to the Kino app.
- The **User Experience** category contains these posts: Autoplay, Picture-in-Picture, Google Cast.
- The **Enhancements** category contains these posts: Encryted Media Extensions, Playback Performance, Background Fetch.
- Boolean flags were implemented to enable certain features only for posts that describe those features.
  - Example: there is a `cast: true` flag first set on the Google Cast post that enables the cast button on that post.

![image](https://user-images.githubusercontent.com/433570/159914180-42b8ef76-bd32-4d8c-a139-4482c13f511a.png)

beaufortfrancois commented on src/api/03-user-experience/07-autoplay.md:
```diff
@@ -0,0 +1,225 @@
+</video>
+<script>
+  const videoElement = document.querySelector('video');
+  videoElement.play();
```
Add `.then(() => { /* video is playing... */ })` so that it's clear it returns a promise.

dero replied:
Updated in 281290836b44cdc445dea0ef7410d1082e9d8506.

beaufortfrancois commented on src/api/03-user-experience/07-autoplay.md:
```diff
@@ -0,0 +1,225 @@
+Similarly to the `autoplay` attribute, browsers may choose to not play the
+video in this situation, especially if it contains sound and is not muted.
+
+One useful benefit of calling the `play()` method is that it returns a Promise
```
Do not capitalize Promise, use `promise`.

dero replied:
Updated in 281290836b44cdc445dea0ef7410d1082e9d8506.

beaufortfrancois commented on src/api/03-user-experience/07-autoplay.md:
```diff
@@ -0,0 +1,225 @@
+
+   /**
+    * Note: In older browsers the `play()` method may not
+    *       return a Promise. If you need to support
```
Same here. Use `promise`

dero replied:
Updated in 281290836b44cdc445dea0ef7410d1082e9d8506.

beaufortfrancois commented on src/api/03-user-experience/07-autoplay.md:
```diff
@@ -0,0 +1,225 @@
+
+## What's Next?
+
+TBD
```
Shall we remove TBD and link to Picture-in-Picture post?

dero replied:
Updated in 281290836b44cdc445dea0ef7410d1082e9d8506.

beaufortfrancois commented on src/api/03-user-experience/08-picture-in-picture.md:
```diff
@@ -0,0 +1,134 @@
+
+## Checking for feature support
+
+[Picture-in-Picture API support] is spotty. Even in browsers that implement it, the feature may be disabled by the `picture-in-picture` Feature Policy. There is a handy `pictureInPictureEnabled` property that will be present on the `document` in browsers that implement the API and its value will be `true` if Picture-in-Picture mode is allowed.
```
replace `is spotty` with something like `is different from what you're probably used to.`

dero replied:
Updated in 281290836b44cdc445dea0ef7410d1082e9d8506.

beaufortfrancois commented on src/api/03-user-experience/08-picture-in-picture.md:
```diff
@@ -0,0 +1,134 @@
+
+## What's next?
+
+Playing videos in a floating window is useful for a local viewing session. However users ofter prefer to watch content on other devices such as their TVs. In the [next article] we'll talk about media casting – and Google Cast in particular.
```
Nit: `s/ofter/often`

dero replied:
Updated in 281290836b44cdc445dea0ef7410d1082e9d8506.

beaufortfrancois commented on src/api/03-user-experience/09-google-cast.md:
```diff
@@ -0,0 +1,247 @@
+
+## Using the Cast Application Framework
+
+The [Cast Application Framework] is the part of Google Cast SDK which exposes objects, methods and events necessary for building web sender applications.
```
You may want to append `(CAF)` to "Cast Application Framework" so that it's clear in the example what CAF is

dero replied:
Updated in 281290836b44cdc445dea0ef7410d1082e9d8506.

beaufortfrancois commented on src/api/03-user-experience/09-google-cast.md:
```diff
@@ -0,0 +1,247 @@
+    e => castVideo(e.sessionState),
+  );
+
+  const castVideo = asnyc (state) => {
```
`s/asnyc/async/`

dero replied:
Updated in 281290836b44cdc445dea0ef7410d1082e9d8506.

beaufortfrancois commented on src/api/03-user-experience/09-google-cast.md:
```diff
@@ -0,0 +1,247 @@
+    if (state === 'SESSION_STARTED') {
+      const session = context.getCurrentSession();
+      const mediaInfo = new chrome.cast.media.MediaInfo(
+        'http://localhost/video.mp4',                // Provide a URL.
```
`s/localhost/example.com`

beaufortfrancois commented on src/api/03-user-experience/09-google-cast.md:
````diff
@@ -0,0 +1,247 @@
+
+```js
+const mediaInfo = new chrome.cast.media.MediaInfo(
+  'http://localhost/video.mp4',                // Provide a URL.
````
`s/localhost/example.com`

dero replied:
Updated in 281290836b44cdc445dea0ef7410d1082e9d8506.

beaufortfrancois commented on src/api/03-user-experience/09-google-cast.md:
```diff
@@ -0,0 +1,247 @@
+);
+
+// Image URLs need to be "wrapped" by the Image class.
+const thumbnail = new chrome.cast.Image('http://localhost/thumb.jpg');
```
`s/localhost/example.com`

dero replied:
Updated in 281290836b44cdc445dea0ef7410d1082e9d8506.

beaufortfrancois commented on src/api/03-user-experience/09-google-cast.md:
````diff
@@ -0,0 +1,247 @@
+mediaInfo.metadata = new chrome.cast.media.GenericMediaMetadata();
+mediaInfo.metadata.title = 'My Dog Chasing Geese';
+mediaInfo.metadata.images = [ thumbnail ];
+```
````
It would be good to remind people `mediaInfo` is used in `chrome.cast.media.LoadRequest(mediaInfo);`

dero replied:
Updated in 281290836b44cdc445dea0ef7410d1082e9d8506.

beaufortfrancois commented on src/api/03-user-experience/09-google-cast.md:
````diff
@@ -0,0 +1,247 @@
+// how and whether additional tracks are exposed in the UI depends
+// on the receiver application implementation.
+mediaInfo.tracks = [ captionsTrack ];
+```
````
It would be good to remind people `mediaInfo` is used in `chrome.cast.media.LoadRequest(mediaInfo);`

dero replied:
Updated in 281290836b44cdc445dea0ef7410d1082e9d8506.

beaufortfrancois commented on src/api/03-user-experience/09-google-cast.md:
````diff
@@ -0,0 +1,247 @@
+
+```js
+const mediaInfo = new chrome.cast.media.MediaInfo(
+  'http://localhost/video.mp4',                // Provide a URL.
````
`s/localhost/example.com`

dero replied:
Updated in 281290836b44cdc445dea0ef7410d1082e9d8506.

beaufortfrancois commented on src/api/03-user-experience/09-google-cast.md:
```diff
@@ -0,0 +1,247 @@
+  chrome.cast.media.TrackType.TEXT,
+);
+
+captionsTrack.trackContentId = 'http://localhost/captions-en.vtt';
```
`s/localhost/example.com`

dero replied:
Updated in 281290836b44cdc445dea0ef7410d1082e9d8506.

beaufortfrancois commented on src/api/03-user-experience/09-google-cast.md:
````diff
@@ -0,0 +1,247 @@
+To achieve this, we can listen for `SESSION_STATE_CHANGED` events and add or remove an overlay element when session is started and ended:
+
+```html
+<style>
````
I'm not sure CSS is necessary in this context as we can use html attribute hidden to hide button

dero replied:
The `.hidden` class is applied to the video overlay that shows a "Casting to [device name]" message on top of the original video element rect.

beaufortfrancois replied:
I was suggesting using `overlay.setAttribute("hidden", "");` and `overlay.removeAttribute("hidden")` in JS code and removes the CSS to make the code sample shorter. WDYT?
I don't feel strongly about it though ;)

beaufortfrancois replied:
Did you remove CSS as well?

beaufortfrancois commented on src/api/03-user-experience/09-google-cast.md:
```diff
@@ -0,0 +1,247 @@
+          const targetName = session.getCastDevice().friendlyName;
+
+          overlay.classList.add('hidden');
+          target.innerText = ` to ${targetName}`;
```
Shall we have "Casting to" and only `targetName` in target to avoid i18n issues?

dero replied:
Right, that would make sense. But, is `session.getCastDevice().friendlyName` always non-empty?

The current implementation should only append the `to [device name]` string to the overlay when device name is not empty. I guess we could sub in something like `external device` in cases when the `friendlyName` is empty.

Right now, when the `friendlyName` is empty, the overlay displays just `Casting`.

beaufortfrancois replied:
I'd assume it's not empty because it is mandatory according to https://developers.google.com/cast/docs/reference/web_sender/chrome.cast.Receiver

dero replied:
Good point. Updated in [e4fb7f7](https://github.com/GoogleChrome/kino/pull/182/commits/e4fb7f7a018355634be2fc0ea61870209c859e3c).

beaufortfrancois commented on src/api/04-enhancements/12-background-fetch-api.md:
```diff
@@ -0,0 +1,166 @@
+    const promises = records.map(async (record) => {
+      const response = await record.responseReady;
+
+      // Maybe store the `Response` object using
```
In this case, we'd want to store those `Response` objects.
Maybe `TODO: Store the response object using the Cache or IndexedDB API`

dero replied:
Updated in 281290836b44cdc445dea0ef7410d1082e9d8506.

beaufortfrancois commented on src/api/04-enhancements/12-background-fetch-api.md:
```diff
@@ -0,0 +1,166 @@
+      // the Cache or IndexedDB API?
+    });
+
+    // Wait until the promises settle.
```
`// Wait until all responses are ready and saved locally`

dero replied:
Updated in 281290836b44cdc445dea0ef7410d1082e9d8506.

beaufortfrancois commented on src/js/utils/decryptVideo.js:
```diff
@@ -0,0 +1,143 @@
+    mediaSource.addEventListener('sourceopen', () => {
+      const sourceBuffer = mediaSource.addSourceBuffer(encryption.mimeCodec);
+
+      fetch(encryption.src).then((response) => {
```
Nit: What happens if `fetch` fails?

dero replied:
Updated in 281290836b44cdc445dea0ef7410d1082e9d8506, added a `catch` block.

beaufortfrancois commented on src/js/utils/decryptVideo.js:
```diff
@@ -0,0 +1,143 @@
+ * Uses Encrypted Media Extensions to decrypt the video.
+ *
+ * @param {HTMLMediaElement} videoElement                    Video element.
+ * @param {object}           encryption                      Encryption data.
```
`encryption.src` is missing in those comments

dero replied:
Good 👀! Updated in 281290836b44cdc445dea0ef7410d1082e9d8506.

beaufortfrancois reviewed (commented):
Great work folks! There is plenty of good material!

beaufortfrancois commented:
Note I'll be OOO next week,

dero commented:
Thank you very much (!) for the detailed feedback here, @beaufortfrancois!

I applied your suggested changes in most cases and only left two conversations open, posting more context and / or questions.

beaufortfrancois commented:
LGTM with nits. Thank you @dero!

dero commented:
Thank you, @beaufortfrancois, I pushed a couple more small fixes addressing the remaining feedback.

Also had the new articles proof-read internally and committed the revised text. Quite a few grammar and style changes, no content changes, though.

If I could get your 👀 for one last time here? We can merge the PR after approval.

beaufortfrancois reviewed (approved):
LGTM % removing unneeded `<style>`

dero commented:
@beaufortfrancois I think the `<style>` is still needed to style the overlay, correct?

But I've removed the now superfluous `.hidden` ruleset. Is that what you had in mind? Thanks for your patience here!

beaufortfrancois commented:
@dero I was thinking of removing entirely the CSS to focus solely on the JS.

```html
<div class="cast-overlay" hidden>
  Casting to <span class="cast-target"></span>
</div>
  
<script>
  // The following code should be defined within
  // `function initCastApi() { ... }` defined previously.
  const context = cast.framework.CastContext.getInstance();

  const overlay = document.querySelector('.cast-overlay');
  const target = document.querySelector('.cast-target');

  // Bind the overlay visibility state to session state changes.
  context.addEventListener(
    cast.framework.CastContextEventType.SESSION_STATE_CHANGED,
    e => {
      switch (e.sessionState) {
        case 'SESSION_ENDED':
          overlay.setAttribute('hidden', '');
          break;
        case 'SESSION_STARTED':
        case 'SESSION_RESUMED':
          const session = context.getCurrentSession();
          target.innerText = session.getCastDevice().friendlyName;
          overlay.removeAttribute('hidden');
      }
    },
  );
</script>
```

beaufortfrancois reviewed (approved):
Thank you for your great work @dero!
